Ticket 120: Hang up call with contact

andioneto opened the pull request:
## Handle ending a call between Twilio client and contact when a call update attempt is made with a status of 'HUNG_UP':
- PUT to /user/:id/campaings/:campaignId/calls/:callId/  [req.body.status = 'HUNG_UP ]
###Added:
- hangUpContactCall [callSid, userId] (in server/util/twilio.js)
- hangUpHandleFlow [res, user_id, call_id, newStatus] ( in server/controllers/calls.js)
-- triggered when the status sent in the request body is HUNG_UP

**TODO:** waiting for ticket 126 to go in so I can do a walkthrough test

andioneto committed:
started on ticket
going to cut new branch and pull in my other work to try to see if I'm getting the response

andioneto committed:
at a stopping point: waiting for 126 and 125 to go in

andioneto committed:
finished up cleaning code

andioneto committed:
updated lint ignore: annoying arrow return rule
clean user in user service wasn't removing hashes from the 'previous attributes'
I wrote a cb handler for the connect update, but that might be more appropriate for Duy's PR
waiting for other code to go in so I can get a better idea of what to expect after the contact-volunteer call is ended ('on hangup'

andioneto committed:
removed change to es lint, I'll talk to team awesome frist

andioneto committed:
removed changes to testing files bc they are being changed drastically... and there's no point yet.

andioneto committed:
linting fixes

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-120-endvolcall

andioneto committed:
integrated hangup flow with the most updated code

broke out hang up flow into a helper function and invoked it inside the controller

andioneto committed:
hold until 126 goes in

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-120-endvolcall

andioneto committed:
successfully able to call, hang up and save call sid!

andioneto committed:
fixing my stupid

andioneto committed:
added a test for the hang up call mutation

perfectlynormalbeast commented on server/controllers/calls.js:
```diff
@@ -140,9 +168,7 @@ export function recordAttempt(req, res) {
-                return callsService.updateCallStatus({ id: call_id, status: newStatus })
-                  .then(updatedCall => res.status(200).json({ message: `call status updated to ${newStatus}`, call: updatedCall }))
-                  .catch(err => console.log('error updating status in calls controller: ', err));
+                return handleHangUpFlow(res, user_id, call_id, newStatus);
```
Thanks for not adding to this bloated method!

andioneto replied:
💃

perfectlynormalbeast commented on server/controllers/calls.js:
```diff
@@ -107,6 +108,33 @@ function afterPutCallAttempt(res, outcome, contact_id, attempt_num, campaign_id)
+  return usersService.getUserById({ id: user_id })
+  .then((userObj) => {
+    const { call_sid: userCallSid } = userObj.attributes;
+    console.log(userObj.attributes);
```
Can we remove these console logs if we're done debugging?

andioneto replied:
Yeah, sorry

perfectlynormalbeast commented on server/controllers/calls.js:
```diff
@@ -107,6 +108,33 @@ function afterPutCallAttempt(res, outcome, contact_id, attempt_num, campaign_id)
   return res.status(200).json({ message: 'call log successfully updated' });
 }
+
+function handleHangUpFlow(res, user_id, call_id, newStatus) {
```
Why do we need to accept a newStatus param when we really just always use status HUNG_UP as the name of the method suggests? I think it'll be cleaner if we remove it.

andioneto replied:
Ture ture, I'll get on it.

andioneto replied:
O.o *true

perfectlynormalbeast commented on server/controllers/calls.js:
```diff
@@ -107,6 +108,33 @@ function afterPutCallAttempt(res, outcome, contact_id, attempt_num, campaign_id)
+    .then(() => {
+      callsService.updateCallStatus({ id: call_id, status: newStatus })
+      .then((updatedCall) => {
+        const updateCallSuccess = `call status updated to ${newStatus}`;
```
We'll be able to get rid of this render and just use HUNG_UP if we remove `newStatus` above.

perfectlynormalbeast commented on server/controllers/calls.js:
```diff
@@ -107,6 +108,33 @@ function afterPutCallAttempt(res, outcome, contact_id, attempt_num, campaign_id)
+      return res.status(500).json({ message: hangUpCallError });
+    });
+  })
+  .catch(err => err);
```
Perhaps return a 404 since we couldn't find the user?

andioneto replied:
Good suggestion, done/

perfectlynormalbeast commented on server/controllers/calls.js:
```diff
@@ -238,3 +264,22 @@ export function hangUpCall(req, res) {
     });
 }
+
+/*
```
Do we need to keep this commented code around?

andioneto replied:
oops, nope, I removed it.

perfectlynormalbeast commented on server/util/twilio.js:
```diff
@@ -44,12 +44,20 @@ export function sayCallCompleted() {
+  return twilioClient.calls(callSid)
+    .update({
+      status: 'completed',
+      statusCallback: `${DEV_PATH}/users/${user_id}/bridge`
```
No - this is not how we're supposed to hang up the call. We hang it up by changing the call's url to the bridge url. Changing the status to completed will actually terminate the call with the volunteer! And we don't want to change the statusCallback either. Does this make sense?

andioneto replied:
hmmm kind of, I'll look back at the docs and see what the other props are I can choose from.

perfectlynormalbeast commented on server/controllers/users.js:
```diff
@@ -280,7 +280,7 @@ export function volunteerCallback(req, res) {
 export function contactCallback(req, res) {
   const { call_id: id } = req.params;
-  const { call_sid } = req.body;
+  const { DialCallSid: call_sid } = req.body;
```
Good catch.

perfectlynormalbeast replied:
Is there a test to be updated for this? I feel like there should be?

andioneto replied:
Yes, this will be in the callback endpoint tests... I think Duy had that ticket, but needed to wait until we connected it to see what the response obj would be.

perfectlynormalbeast commented on server/util/twilio.js:
```diff
@@ -44,12 +44,20 @@ export function sayCallCompleted() {
   helloContact.say(`Now dialing ${contactName}`);
-  helloContact.dial({ action: `${DEV_PATH}/user/${userId}/campaign/${campaignId}/calls/${callId}/callback`, method: 'POST' }, contactNumber);
-  helloContact.redirect({ method: 'POST' }, `${DEV_PATH}/user/${userId}/campaign/${campaignId}/calls/bridge`);
+  helloContact.dial({ action: `${DEV_PATH}/users/${userId}/campaigns/${campaignId}/calls/${callId}/callback`, method: 'POST' }, contactNumber);
```
again, good catch!

perfectlynormalbeast commented on test/server/util/twilio_mocks.js:
```diff
@@ -4,6 +4,7 @@ const { TWILIO_ACCOUNT_SID } = process.env;
 
 export const CALL_COMPLETE = 'CALL_COMPLETE';
 export const CONNECT_CONTACT_URL = 'CONNECT_CONTACT_URL';
+export const DISCNNECT_CONTACT_URL = 'DISCNNECT_CONTACT_URL';
```
Typo?

andioneto replied:
lol yes.

perfectlynormalbeast commented on test/server/util/twilio_mocks.js:
```diff
@@ -35,15 +36,20 @@ export function returnUpdate(callSid, updateType) {
+      clientUpdateObject.status = 'in-progress';
+      return clientUpdateObject;
+    case DISCNNECT_CONTACT_URL:
+      clientUpdateObject.status = 'completed';
```
So yeah we don't actually terminate the volunteer call. So we'll have to modify this test too. There really won't be much that changes in the call object IIRC since we're transparently changing the URL which isn't reflected in the response from the Twilio API or client.

andioneto replied:
hmm ok, I might have a few questions about this in a bit

perfectlynormalbeast commented on test/server/util/twilio_mocks.js:
```diff
@@ -54,19 +60,26 @@ export function createCall() {
       update: (options) => {
         const optionsCalled = Object.getOwnPropertyNames(options).sort();
-        if (lodash.isEqual(optionsCalled, [ 'status', 'statusCallback' ])) {
+        if (lodash.isEqual(optionsCalled, ['status', 'statusCallback'])) {
```
We'll have to change this logic in the mock as well since we're not specifying a new statusCallback url for the terminating the contact call.

andioneto replied:
Yeah, that actually will make this test mock helper a bit cleaner, the function I wrote to distinguish between the two .update calls was a bit hacky.

perfectlynormalbeast reviewed (changes requested):
So this PR doesn't actually end the contact call like we discussed earlier. Hope the comments help. If not, we can chat.

andioneto committed:
made requested changes
also removed test since the call  mutation doesn't change the twilio response at all

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-120-endvolcall

perfectlynormalbeast commented on server/util/twilio.js:
```diff
@@ -68,3 +61,10 @@ export function mutateCallConnectContact(callSid, idCollection) {
+export function hangUpContactCall(callSid, userId, campaignId) {
+  return twilioClient.calls(callSid)
+    .update({
+      url: `${DEV_PATH}/users/${userId}/campaigns/${campaignId}/calls/bridge`
```
Hope we've confirmed this is where the bridge Twiml lives? Should be identical to the redirect at the end of the call twiml.

perfectlynormalbeast commented on test/server/util/twilio.js:
```diff
@@ -18,14 +18,7 @@ describe('Twilio client methods', function() {
-    done();
-  });
-  // test for hang up
+  // test for hang up ?
```
There really isn't a good way to test this other than asserting that we get a call object back from Twilio. Maybe a timestamp changes but that's not a good way to confirm in a unit test.

perfectlynormalbeast reviewed (dismissed):
+1 one with one comment? also conflicts?

andioneto committed:
added some minor changes to fix the call status update on hangup RESOLVED

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-120-endvolcall

andioneto committed:
Merge branch 'master' into ao-120-endvolcall
